Automattic/wp-calypso
# Change Scan review-prompt to display after 3 visits to Scan page.

### Changes proposed in this Pull Request

This PR updates the Jetpack-Review-Prompt for Jetpack Scan to display after 3 visits to the Scan page (not on the same day), instead of after a Scan threat is successfully "auto-fixed". 

Related to: 1164141197617539-as-1201731270035137

**Jetpack Review Prompt context:**
The Jetpack Review Prompt is a UI component that displays to users, prompting them to leave a review for Jetpack on the WordPress plugin repository.  There is a review prompt for Backup and for Scan. The Backup one displays after the user initiates a site restore. The Scan one, prior to this PR, displays after the user successfully "auto-fixes" a detected scan threat.
The way the Review Prompts dismiss behavior works is: After the prompt has been triggered to display, the user can dismiss the prompt by either clicking the (X) button in the top right corner, or by clicking the "Leave a review" button. If the "Leave a review" button is clicked, the prompt is dismissed permanently. If the prompt is dismissed using the (X) button, the prompt will dismiss initially, but will appear again in 2 weeks. When the prompt is dismissed the second time, it will then dismiss permanently.

Instead of triggering the Scan Review Prompt after a scan threat has been auto-fixed, this PR will trigger the Scan Review prompt after 3 visits (on separate days) to the main Scan page.



**Additional implementation notes:**
- Created a new "persistent-counter" [preference](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/preferences/README.md), and uses this to _persistently_ count each Scan page visit. I tried to make the persistent-counter extensible so it can be reused for persistently counting any other events or actions within Calypso, if needed.
- Created unit tests for the "persistent-counter".
- Updated the existing Jetpack Review Prompt (for Scan) to display after 3 visits (on different days) to the main Scan page, instead of after a scan threat is auto-fixed.
- Updated the existing Jetpack Review Prompt (for Scan) to **no longer** store the review prompt preferences values **per each user's site**. This is no longer needed because the review prompt is no longer based on the scan threat auto-fix of _each site_.

### Testing instructions

1. Spin up this PR in Calypso green (`yarn start-jetpack-cloud`). (or green and blue concurrently)
1. Open up file `client/my-sites/scan/main.tsx`, line 79, and change `dispatchIncrementCounter( SCAN_VISIT_COUNTER_NAME, false, false );` to:
```
dispatchIncrementCounter( SCAN_VISIT_COUNTER_NAME, false, true ); // last argument, set to true.
```
(This change will allow the page counter to count on every visit (for testing purposes), instead of only counting once per day).
1. Select a Jetpack site that has an active Scan subscription (or purchase a new Scan product)
1. Visit `http://jetpack.cloud.localhost:3000/scan/:site` (replace :site with your site slug)
1. If interested you can open Redux DevTools, and inspect the global state object. Go to: preferences -> remoteValues -> persistent-counter -> scan-page-visit, to view the counter increment on every scan page visit.
1. <img width="614" alt="Markup 2022-03-10 at 17 08 41" src="https://user-images.githubusercontent.com/11078128/157774342-e16c31bf-7191-4d54-9c21-076604631afc.png">
1. Reload the scan page (or navigate away and back again) _2 more times_. Upon **3 total visits** you should see the Jetpack Review Prompt (see screenshot):
1. ![Markup 2022-03-10 at 17 12 42](https://user-images.githubusercontent.com/11078128/157774386-55855377-173d-4444-b3d3-aacee77b9a56.png)
1. Dismiss the Jetpack Review Prompt by clicking the (X) close button in the top right corner. The prompt should disappear.
1. Now set your computer's clock forward 2+ weeks. For Mac go to System Preferences -> Date & Time -> Manually set date forward 2 weeks and 1 day (15 days) or more.
1. Reload the scan page, you should see the Jetpack Review Prompt again.
1. Dismiss the Review Prompt again clicking the (X) close button. The prompt should disappear.
1. Manually set your computer clock again to an additional 15 or more days forward.
1. Reload the scan page again and verify the Review Prompt **does not** appear again.
1. And that's it.
1. Reset your computer's clock back to the correct date.
1. To reset the counter and review prompt Preferences settings back to default, Open up the Environment badge thingy in the lower right corner: 
![Markup 2022-03-10 at 17 58 37](https://user-images.githubusercontent.com/11078128/157774586-2519a593-b673-45dc-920d-f7776e91dd78.png)
1. hover over "Preferences" and click the (X) button on the "persistent-counter" and "jetpack-review-prompt" preferences properties.  This will delete/reset the counter and review-prompt values. (See video below).

https://user-images.githubusercontent.com/11078128/157774619-da48eacc-447a-49dd-b2ed-77b584ba934c.mp4

#### Run unit tests:

Run:

` yarn test-client client/state/persistent-counter`

`yarn test-client client/state/jetpack-review-prompt`
